burn-wgpu/src/context.rs

```diff
@@ -0,0 +1,252 @@
+use burn_common::id::IdGenerator;
+use std::{
+    any::TypeId,
+    borrow::Cow,
+    collections::HashMap,
+    sync::{Arc, Mutex},
+};
+
+use wgpu::{
+    util::{BufferInitDescriptor, DeviceExt},
+    Buffer, DeviceDescriptor, DeviceType, ShaderModule, ShaderModuleDescriptor,
+};
+
+use crate::{kernel::KernelGenerator, GraphicsAPI, WGPUDevice};
+
+/// The context is the basic struct that allows to execute GPU kernel on devices.
+///
+/// You can access a context for a [wgpu device](WGPUDevice) using [get_context](crate::pool::get_context).
+#[derive(Debug)]
+pub struct Context {
+    id: String,
+    queue: wgpu::Queue,
+    device_wgpu: wgpu::Device,
+    cache: Mutex<HashMap<TypeId, Arc<ShaderModule>>>,
+    pub(crate) device: WGPUDevice,
+}
+
+#[derive(new, Clone, Debug)]
+pub struct WorkGroup {
+    pub x: u32,
+    pub y: u32,
+    pub z: u32,
+}
+
+impl Context {
+    pub(crate) fn new<G: GraphicsAPI>(device: &WGPUDevice) -> Self {
+        // Instantiates instance of WebGPU
+        let instance = wgpu::Instance::default();
+
+        // `request_adapter` instantiates the general connection to the GPU
+        let adapters = instance.enumerate_adapters(G::backend().into());
+        let mut adapters = adapters
+            .filter(|adapter| {
+                let device_type = adapter.get_info().device_type;
+                match device {
+                    WGPUDevice::DiscreteGPU(_) => device_type == DeviceType::DiscreteGpu,
+                    WGPUDevice::IntegratedGPU(_) => device_type == DeviceType::IntegratedGpu,
+                    WGPUDevice::VirtualGPU(_) => device_type == DeviceType::VirtualGpu,
+                    WGPUDevice::CPU => device_type == DeviceType::Cpu,
+                }
+            })
+            .collect::<Vec<_>>();
+
+        let adapter = match device {
+            WGPUDevice::DiscreteGPU(num) => {
+                assert!(adapters.len() > *num, "No Discrete GPU device found");
+                adapters.remove(*num)
+            }
+            WGPUDevice::IntegratedGPU(num) => {
+                assert!(adapters.len() > *num, "No Integrated GPU device found");
+                adapters.remove(*num)
+            }
+            WGPUDevice::VirtualGPU(num) => {
+                assert!(adapters.len() > *num, "No Virtual GPU device found");
+                adapters.remove(*num)
+            }
+            WGPUDevice::CPU => {
+                assert!(!adapters.is_empty(), "No CPU device found");
+                adapters.remove(0)
+            }
+        };
+
+        let device_wgpu = device.clone();
+        let (device, queue) = pollster::block_on(adapter.request_device(
+            &DeviceDescriptor {
+                label: None,
+                features: wgpu::Features::empty(),
+                limits: wgpu::Limits::downlevel_defaults(),
+            },
+            None,
+        ))
+        .unwrap();
```

Comment:
Probably we should make it clearer with expect message. I am not sure what would failing here.